subsys/bluetooth/host/conn.c

```diff
@@ -1181,11 +1196,21 @@ void bt_conn_notify_tx(struct bt_conn *conn)
 
 	while ((tx = k_fifo_get(&conn->tx_notify, K_NO_WAIT))) {
 		BT_DBG("cb %p user_data %p", tx->data.cb, tx->data.user_data);
+
+#if defined(CONFIG_BT_CONN_DEFER_TX_CB)
+		/* Only submit if there is a callback set */
+		if (tx->data.cb) {
+			k_work_submit(&tx->work);
```

Comment:
I fear this might cause deadlocks. IIRC the way this was designed is that we do the TX callbacks in the TX thread *before* trying to send new packets to the controller, so that when the thread comes to the point of sending more packets it's guaranteed to have the necessary TX contexts available (there's e.g. a code comment about this in the send_frag function)

Reply:
Actually, not a deadlock, but there's an `__ASSERT()` that'll be triggered in `add_pending_tx()` if/when this happens since this allocation is by design supposed to always succeed.

Reply:
Afaik the risk of the deadlock is already present in the current form since the callback can block as well, though I guess you are saying that when defering we need to make ensure there are enough context to continue since they are no longer freed immediately which I guess can be resolved by using k_fifo for free_tx so it can wait with k_fifo_get instead of asserting.

Reply:
> Afaik the risk of the deadlock is already present in the current form since the callback can block as well

@Vudentz no, since this is in the same thread, i.e. currently the allocation part doesn't happen until the callback returns.

> though I guess you are saying that when defering we need to make ensure there are enough context to continue since they are no longer freed immediately which I guess can be resolved by using k_fifo for free_tx so it can wait with k_fifo_get instead of asserting.

Basically this thread runs whenever there's a HCI Number of Completed Packets event. Each "packet" maps to a tx context, so the idea is that this context is freed up for reuse by the HCI event. Using a k_fifo for the tx contexts and then blocking in the TX thread would indeed work, but then we need a k_fifo (which takes more memory) and we also need to extend the tx context with 4 bytes for the k_fifo (which also takes more memory), so either way we're looking at increasing the memory footprint. Making the TX thread size user configurable is at least complexity-wise the simpler option since that doesn't require adding any new code.

Reply:
> > though I guess you are saying that when defering we need to make ensure there are enough context to continue since they are no longer freed immediately which I guess can be resolved by using k_fifo for free_tx so it can wait with k_fifo_get instead of asserting.
> 
> Basically this thread runs whenever there's a HCI Number of Completed Packets event. Each "packet" maps to a tx context, so the idea is that this context is freed up for reuse by the HCI event. Using a k_fifo for the tx contexts and then blocking in the TX thread would indeed work, but then we need a k_fifo (which takes more memory) and we also need to extend the tx context with 4 bytes for the k_fifo (which also takes more memory), so either way we're looking at increasing the memory footprint. Making the TX thread size user configurable is at least complexity-wise the simpler option since that doesn't require adding any new code.

So I did the change to k_fifo which essencially uses the same node, since it is either on tx_free or tx_pending, I think this is the safest bet since we don't incor in using more memory and have the safety of using k_fifo_get which does properly wait for the callbacks to complete.

Reply:
Btw, it seems I one would have substantially increase the stack size in order to have something like a shell_print on the callback, which is why I continue with system wq approach.

Anyway the use of k_fifo_get is propably safer even when not defering so even if we were to drop the use of system wq I think it best to keep the use of k_fifo for tx_free.

Reply:
Yes, makes sense. If forgot the conn_tx contexts already had the necessary space for being inserted into a k_fifo.